### ISSUE TITLE
Recreating custom element loads wrong route

### ISSUE THREAD

**Issue by jsBrian** [angular/angular#36711]
# 🐞 bug report

### Affected Package
@angular/elements
@angular/router

### Is this a regression?
No, as far is I know this issue was present in @angular/elements 6, 7, 8 and 9.

### Description
When using custom-elements along with routing, there seems to be an issue when you:
* create a custom element (that has routing) and add it to the page
* do some routing, but make sure to back to the root of the app
* delete the custom element
* create a new element with the same tag name
* instead of showing the root of the app, it opens one of the child routes and routing breaks

The minimal reproduction should make the issue much more clear.

## 🔬 Minimal Reproduction
https://stackblitz.com/edit/angular-elements-official-example-r3ipbj

## 🔥 Exception or Error
<pre><code>No error is present</code></pre>

**Anything else relevant?**
The issue also seem to arise when you reuse the old custom element. But only when you remove and add it twice.

**Comment by jsBrian** [comment]
tl;dr the router-outlet reactivates itself when recreating an element, even when it's not supposed to

Performed some tests and found out some more information.
It seems that the old router-outlet gets reactivated when recreating the element. By manually deactivating the router-outlet, if it's not expected, in the ngAfterViewInit call, I can bypass the issue.

The reason routing broke is that the primary router outlet is active, but it's not supposed to be. When trying to navigate the router then tries to activate the router-outlet (as in its internal state, it's not activated yet). However, the router-outlet is already activated.
Normally this would throw an error, but somehow this error isn't thrown either.

**Comment by Wackymax** [comment]
> tl;dr the router-outlet reactivates itself when recreating an element, even when it's not supposed to
> 
> Performed some tests and found out some more information. It seems that the old router-outlet gets reactivated when recreating the element. By manually deactivating the router-outlet, if it's not expected, in the ngAfterViewInit call, I can bypass the issue.
> 
> The reason routing broke is that the primary router outlet is active, but it's not supposed to be. When trying to navigate the router then tries to activate the router-outlet (as in its internal state, it's not activated yet). However, the router-outlet is already activated. Normally this would throw an error, but somehow this error isn't thrown either.

I am trying to implement your bypass but I am struggling to figure out how you did it. Can perhaps give an example that myself and others running into this issue can use?

**Comment by atscott** [comment]
The issue here (after updating the example to v14) is that the new `RouterOutlet` instance is getting created and initialized before the `ngOnDestroy` runs for the one being removed. If I put the creation and insertion of the new element inside a `setTimeout(() => {...}, 10)` to ensure correct ordering, then the issue is resolved. 

If the `ngOnDestroy` were to pass the outlet instance here https://github.com/angular/angular/blob/0934011b03aae6a4848ab4eeacc896061f4dceca/packages/router/src/directives/router_outlet.ts#L186 and the context did not clear the outlet if the instance is already set to a different one, then I believe that would resolve the issue in Angular as well. While this seems to make sense, I'm a bit hesitant to implement this fix without more of a signal that it should be done and is correct.

**Comment by atscott** [comment]
After further consideration, I'm going to close this. The `RouterOutlet` is not made to support multiple registered instances for the same outlet name, and this is what's happening. Baking in a "fix" for these out of order register/deregister operations would just be a hack for an unsupported use-case. 

Instead, you should either ensure that you aren't creating a second outlet before the first is destroyed, or create a custom router outlet implementation that can handle this. For example, copy-paste the current implementation and change the `ngOnDestroy` to 

```
  ngOnDestroy(): void {
    // Ensure that the registered outlet is this one before removing it on the context.
    if (this.parentContexts.getContext(this.name)?.outlet === this) {
      this.parentContexts.onChildOutletDestroyed(this.name);
    }
  }
```

Yet another option would be to make the custom implementation actually support multiple `RouterOutlet` instances as described in: https://github.com/angular/angular/issues/31587#issuecomment-660378081

**Comment by atscott** [comment]
Looks like the same timing issue is happening in https://github.com/angular/angular/issues/32453. Since this isn't an isolated use-case anymore and the layout of the application described there is normal, I'll address this directly in the `RouterOutlet`. The issue there is that the constructor where the outlet gets registered is called before the lifecycle hooks `ngOnDestroy` and `ngOnInit` so even though the `ngOnDestroy` of the old one is called first, the new one is already registered.
